### PR TITLE
improve handling of the escape character

### LINKED ISSUE #1356: Support Escape as meta
I am using fish 2.1.0 and Escape is not considered meta.

This is really annoying, especially on OSX. You could configure iTerm to make 'alt' behave like meta, however, and I quote fish website,
> configuration is source of all evil.

Bash, zsh, emacs, and many more support out of the box Escape as meta, which allow a proper usage without having to temper with the terminal configuration.

I understand why Fish is not POSIX, however, I don't see any argument why this should not be supported.

After discussing on IRC, it seems that it IS supported. However, it does not work on OSX with terminal, iTerm2 or with direct TTY within virtualbox.

Is it a bug? Is it expected? If yes, how could I work around it to make it work?

And please, do not tell me to configure my iTerm. I want to use `Escape`, not `Alt`.

### PR DESCRIPTION
**Pull request by krader1961**
Increase the delay between seeing an escape character and giving up on
whether additional characters that match a key binding are seen. I'm
setting the value to 500 ms to match the readline library. We don't need
such a large window for sequences transmitted by a terminal (even over ssh
where network delays can be a problem).  However, we can't expect humans to
reliably press the escape key followed by another key with an inter-char
delay of less than ~250 ms based on my testing and research. A value of
500 ms provides a nice experience even for people using "fish_vi_mode"
bindings as a half second to switch from insert to normal mode is still
fast enough that most people won't even notice.

Resolves #1356

### CODE REVIEW

**Comment by faho** [comment]
Predictably, this causes the binding tests to fail because those don't wait that long. They would have to be adjusted.

I also believe this is something that warrants wider testing. The change itself is trivial (and has the added benefit of being like readline), but the behavioral change is not something I at least can confidently judge right now.

**Comment by krader1961** [comment]
I agree that this should not be merged until we've heard from several people or a month has gone by with no reports of problems. I'll augment the PR to fix the unit tests.

**Comment by krader1961** [comment]
By the way, I didn't notice that the .../tests/test.fish script I was running to verify my changes didn't break anything was insufficient until this issue. The documentation for how to execute the unit tests needs to be augmented. It might also be a good idea to make running all the tests on a local system easier. I'll see about improving both of those aspects of this project.

**Comment by krader1961** [comment]
FWIW, I personally prefer a timeout of 300 ms. But that's mostly because fish currently has a somewhat weird behavior of inheriting the default bindings while in vi-mode. Which means, for example, that \ed is a valid sequence in vi-mode. However, that short of an escape timeout can be a problem for some people.

**Comment by krader1961** [comment]
Okay, I've implemented the ability to configure the escape timeout and augmented the unit tests. Please take another look. I'll add another change to this PR in to update the documentation soon.

**Comment by krader1961** [comment]
According to travis my change failed the tests due to code completely unrelated to my change. WTF?

Testing history race conditions
Error: Line '8939 154' found in history not found in some array
Error: Line '8942 132' found in history not found in some array
Error: Line '8942 131' found in history not found in some array
Error: Line '8942 130' found in history not found in some array
Error: Line '8942 129' found in history not found in some array
Testing history_sample_fish_1_x
Testing history_sample_fish_2_0
Testing bash import
Testing history_sample_corrupt1
Encountered 5 errors in low-level tests
make: *** [test_low_level] Error 1
The command "make test SHOW_INTERACTIVE_LOG=1" exited with 2.

File: doc_src/bind.txt [hunk 1]
```diff
@@ -131,3 +131,14 @@ set -g fish_key_bindings fish_vi_key_bindings
 bind -M insert \cc kill-whole-line force-repaint
 \endfish
 Turns on Vi key bindings and rebinds @key{Control,C} to clear the input line.
+
+
+\subsection special-case-escape Special Case: The escape Character
+
+The escape key (or character) poses a special challenge for fish. Consider Vi mode. On the one hand you want to be able to press the escape key to switch from insert to normal mode. On the other hand you want fish to recognize multi-char sequences that begin with an escape character (which is what the function and arrow-keys send) without having the escape character switch to normal mode. So fish needs to wait a little bit to see if the escape introduces a multi-char sequence but not so long that you're left wondering if fish will ever switch to Vi normal mode after you press escape.
```
Comment (oranja) [inline doc_src/bind.txt]: Even though it is not unprecedented in Fish's command reference to use direct speech ("you"), I would consider rephrasing the majority of this text to use indirect speech. It sounds more appropriate to me (maybe I'm too old fashioned? what do you think?).

Following is a proposal with mostly indirect speech. This is actually more than slight rephrasing, but it's not as if to say that I think it's better now than before. My text is simply different (and worse, if anything).

> Since the escape key (or character) plays two distinct roles, the escape key poses a special challenge for fish. In its first role, it stands by itself. In Vi mode, for example, escape is used to switch from insert to normal mode. In its second role, however, escape is used as a "meta" key and then it's only the beginning of some longer key-sequence. Coming back to the Vi mode example, sometimes fish is expected to realize that the escape key should be regarded as "meta", meaning that the escape character is part of some multi-char sequence. Function keys (F1, F2, etc...) and arrow keys are common cases of multi-char sequences sent with the escape character as "meta", but escape can be used in this manner to define custom bindings as well. Obviously, fish is not supposed to exit insert mode when the escape is part of a longer key-sequence.

> To be able distinguish between these two roles, fish has to wait after it sees an escape character. In this waiting period, any additional key presses make the escape key behave as "meta". Otherwise, it remains simply an escape key press. The waiting period is set to 500 milliseconds (0.5 seconds) by default This is the Gnu readline library default escape timeout. It can be configured by setting the `fish_escape_delay_ms` variable to a value between 10 and 5000 ms. It is recommended that this be a universal variable that you set once from an interactive session.

(The note on previous versions remains unchanged)

The rest of the PR looks great to me, excluding test code, which I am not familiar with and didn't have time to go over.

Comment (krader1961) [inline doc_src/bind.txt]: In general the active (direct) void is preferable to the passive (indirect) voice. Technical literature is especially prone to overuse of the passive voice. Having said that I like what you wrote and will adopt it with a couple of minor changes.

**Comment by krader1961** [comment]
Ping? I think this PR is now mature enough to be merged. Several people have reported a positive experience and no one has reported a negative experience. I also, eventually, managed to get Travis-CI to consistently succeed running the relevant unit test.

**Comment by oranja** [comment]
I'm very much in favor of this PR.
I tried to review the changes, but I didn't have enough time to go over the tests and I honestly don't feel competent enough to make a valid assessment. FWIW, what I've seen, the main changes, looked perfectly fine to me.

@krader1961 I don't know what's the policy here, but perhaps it would be better in the long term to squash this into a few constructive commits and discard the trial-and-error ones? (I am more querying than suggesting.)

**Comment by faho** [comment]
@oranja: What we usually do is that whoever merges it squashes it together into however commits make sense. In this case probably one or two.

@krader1961: Do you know of anyone who has tested it on vi-bindings? I could try of course, but I could never get used to shell vi-bindings, so it wouldn't quite be a fair test.

**Comment by krader1961** [comment]
@oranja: In hindsight I wish I had created a separate branch while figuring out why Travis-CI was failing yet running the tests on my OS X and Ubuntu systems always passed.

@faho: The only person I can vouch has tested this with vi-bindings is myself. As I noted earlier I find the 500 ms default just a trifle annoying and am using "set -U fish_escape_delay_ms 200". The only question  is whether the default timeout should be based on the bindings that are in effect. I'd prefer to not add that complexity, to both the implementation and documentation, since it is trivial for someone to set their desired timeout.

**Comment by faho** [comment]
> I'd prefer to not add that complexity, to both the implementation and documentation, since it is trivial for someone to set their desired timeout.

Humm... it's also not too difficult to add - `set -q fish_escape_delay_ms; or set -g fish_escape_delay_ms 200` in `fish_vi_mode`. We currently don't have a `fish_default_mode` function to return to default keybindings, though, so unsetting vi-mode would leave that delay, but then again it wouldn't unset the mode-indicator either, so it appears nobody has cared enough about that scenario until now.

As I said though I don't use vi-mode so if you're not convinced leave it.

**Comment by krader1961** [comment]
> We currently don't have a fish_default_mode function to return to default keybindings, though, so unsetting vi-mode would leave that delay, 

Yeah, that's the aspect of the implementation I was alluding to. On the other hand, how often do we think people switch between the modes (excluding developers who are testing things)? It could be a non-issue. Too, my gut tells me most users of vi-bindings will want an escape delay closer to 250 to 300ms. If we were doing this at a company like Google we would conduct formal user-interface testing. It's not clear to me how to do anything even remotely like that for this type of project. So do we want to rely on people reading the documentation to find out how to tweak the timeout if they don't like 500ms or go with my gut and have two defaults?

**Comment by faho** [comment]
>If we were doing this at a company like Google we would conduct formal user-interface testing. It's not clear to me how to do anything even remotely like that for this type of project.

I have thought about getting some users to form a focus group, but that doesn't quite seem feasible (how'd I contact them? Plus this'd be experienced users while often you'd want newbies). Also having a group of people to do my bidding might trigger my already apparent super-villainy.

What we can do is guess by looking at the documentation. We need two things:

- Where would a user find out about vi-mode (from our documentation since we can't control blog posts by other people etc)?

- Where would a user find out about escape_delay?

For the former, we have the documentation for `fish_vi_mode` itself, and the section "Command line editor" - which actually recommends calling "fish_vi_key_bindings". And sure enough, github code search shows [146 matches for fish_vi_mode](https://github.com/search?q=fish_vi_mode&type=Code&utf8=%E2%9C%93) and [263 matches for fish_vi_key_bindings](https://github.com/search?q=fish_vi_key_bindings&type=Code&utf8=%E2%9C%93). The latter users won't get the adjustment even if we added it to fish_vi_mode. Also, the documentation doesn't differentiate the two functions well enough - fish_vi_key_bindings just sets the binds, while fish_vi_mode will also set up the rest of the environment (like the prompt).

For the latter, we can just check this PR, and see that it is at the end of the (already comparetively long) bind doc, and in the chapter "Special Variables" that I've forgotten existed.

In other words:

It doesn't really seem like most people will see it, but more people than not might not even get the automatic change if we add it.

**Comment by krader1961** [comment]
@faho: I think it is sufficient to set the Vi mode default escape timeout in _ fish_vi_key_bindings_ since that is called from _fish_vi_mode_. I'll update this PR with that change.

**Comment by ridiculousfish** [comment]
Finally groked this and overall it's a really nice fix. fish is opposed to configuration on principle, but this isn't a case of disagreement so much as reflecting differences in terminal configurations, or even keyboard hardware. I can think of no way to detect the right behavior automatically, so configuration is justified here.

Two thoughts:

1. Dropping the delay in vi mode is a great idea. We can say that that escape-as-meta is unsupported by default in fish_vi_mode, though you can configure it using the variable.

2. 500 ms feels awfully long. Escape-as-meta is an advanced feature, and (speculating) most escape-as-meta users are fast typists. Also non-EaM users will be negatively impacted by a long delay, because escape is used to e.g. clear the pager. I think we should lower this to a level where it feels comfortable for both EaM and otherwise. Users who require a longer delay will have that ability, but we ought to tune for most users who will just use escape as escape. 300 ms seems about right to me. What do you think?

**Comment by krader1961** [comment]
> We can say that that escape-as-meta is unsupported by default in fish_vi_mode,...

That's an interesting idea but I wouldn't say it's unsupported. I'll probably opt for verbiage to the effect that "if you want to be able to use the escape key as a meta key in vi-mode you'll need to increase the escape delay timeout." Then set the default for vi-mode to the 10ms it currently is set to.

> non-EaM users will be negatively impacted by a long delay, because escape is used to e.g. clear the pager...

I don't grok this. Does a bare escape in emacs mode do something useful? Regardless, I'm inclined to agree that setting the default to 300ms is probably closer to what the vast majority of escape-as-meta users need so I'll do that. I only chose 500ms because it's the Gnu readline default.

**Comment by ridiculousfish** [comment]
> Does a bare escape in emacs mode do something useful

I thought it did...looks like the only thing it does is end paging mode (e.g. the tab menu).

**Comment by ridiculousfish** [comment]
Merged as 25ad8866c97b242911b1e52298c8cfeed4530097 with a lot of git surgery. Thanks for taking this on!!
